server/src/compression.rs

```diff
@@ -0,0 +1,473 @@
+}
+
+impl<'a> CompressedSegmentBuilder<'a> {
+    fn new(
+        uncompressed_timestamps: &'a TimestampArray,
+        uncompressed_values: &'a ValueArray,
+        error_bound: ErrorBound,
+    ) -> Self {
+        Self {
+            uncompressed_timestamps,
+            uncompressed_values,
+            start_index: 0,
+            pmc_mean: PMCMean::new(error_bound),
+            pmc_mean_has_fit_all: true,
+            swing: Swing::new(error_bound),
+            swing_has_fit_all: true,
+            gorilla: Gorilla::new(),
+        }
+    }
+
+    /// Attempt to update the current models to also represent the `value` of
+    /// the data point collected at `timestamp`.
+    fn try_to_update_models(&mut self, timestamp: Timestamp, value: Value) {
+        self.pmc_mean_has_fit_all = self.pmc_mean_has_fit_all && self.pmc_mean.fit_value(value);
+
+        self.swing_has_fit_all =
+            self.swing_has_fit_all && self.swing.fit_data_point(timestamp, value);
+
+        // Gorilla uses lossless compression and cannot exceed the error bound.
+        if self.gorilla.length < GORILLA_MAXIMUM_LENGTH {
+            self.gorilla.compress_value(value);
+        }
+    }
+
+    /// Return `true` if any of the current models can represent additional
+    /// values, otherwise `false`.
+    fn can_fit_more(&self) -> bool {
+        self.pmc_mean_has_fit_all
+            || self.swing_has_fit_all
+            || self.gorilla.length < GORILLA_MAXIMUM_LENGTH
+    }
+
+    /// Store the model that requires the smallest number of bits per value in
+    /// `compressed_record_batch_builder`. Returns the index of the value in
+    /// `uncompressed_values` the selected model could not represent.
+    fn finish(self, compressed_record_batch_builder: &mut CompressedRecordBatchBuilder) -> usize {
+        // The model that uses the fewest number of bytes per value is selected.
+        let SelectedModel {
+            model_type_id,
+            end_index,
+            min_value,
+            max_value,
+            values,
+        } = models::select_model(
+            self.start_index,
+            self.pmc_mean,
+            self.swing,
+            self.gorilla,
+            self.uncompressed_values,
+        );
+
+        // Add timestamps and error.
+        let start_time = self.uncompressed_timestamps.value(self.start_index);
+        let end_time = self.uncompressed_timestamps.value(end_index);
+        let timestamps = &[]; // TODO: compress irregular timestamps.
+        let error = f32::NAN; // TODO: compute and store the actual error.
+
+        compressed_record_batch_builder.append_compressed_segment(
+            model_type_id,
+            timestamps,
+            start_time,
+            end_time,
+            &values,
+            min_value,
+            max_value,
+            error,
+        );
+        end_index + 1
+    }
+}
+
+/// A batch of compressed segments being build.
+struct CompressedRecordBatchBuilder {
```

Comment:
Including the type name `Recordbatch` in the name is maybe not ideal. Maybe change the name to `CompressedSegmentBatchBuilder` or something similar.

Reply:
I do not really see a problem with using `RecordBatch` as part of the name, but I have changed it to `CompressedSegmentBatchBuilder`.